src/Dictionaries/HashedDictionary.cpp

```diff
@@ -56,8 +197,62 @@ HashedDictionary<dictionary_key_type, sparse>::HashedDictionary(
     calculateBytesAllocated();
 }
 
-template <DictionaryKeyType dictionary_key_type, bool sparse>
-ColumnPtr HashedDictionary<dictionary_key_type, sparse>::getColumn(
+template <DictionaryKeyType dictionary_key_type, bool sparse, bool sharded>
+HashedDictionary<dictionary_key_type, sparse, sharded>::~HashedDictionary()
+{
+    size_t shards = std::max<size_t>(configuration.shards, 1);
+    size_t attributes_tables = std::max<size_t>(attributes.size(), 1 /* no_attributes_containers */);
+    ThreadPool pool(shards * attributes_tables);
+
+    size_t hash_tables_count = 0;
+    auto schedule_destroy = [&hash_tables_count, &pool](auto & container)
+    {
+        if (container.empty())
+            return;
+
+        pool.scheduleOrThrowOnError([&container, thread_group = CurrentThread::getGroup()]
+        {
+            if (thread_group)
+                CurrentThread::attachToIfDetached(thread_group);
+            setThreadName("HashedDictDtor");
+
+            if constexpr (sparse)
+                container.clear();
+            else
+                container.clearAndShrink();
+        });
+
+        ++hash_tables_count;
+    };
+
+    if (attributes.empty())
+    {
+        for (size_t shard = 0; shard < shards; ++shard)
+        {
+            schedule_destroy(no_attributes_containers[shard]);
+        }
+    }
+    else
+    {
+        for (size_t attribute_index = 0; attribute_index < attributes.size(); ++attribute_index)
+        {
+            getAttributeContainer(attribute_index, [&](auto & containers)
+            {
+                for (size_t shard = 0; shard < shards; ++shard)
+                {
+                    schedule_destroy(containers[shard]);
+                }
+            });
+        }
+    }
+
+    LOG_TRACE(log, "Destroying {} non empty hash tables (using {} threads)", hash_tables_count, pool.getMaxThreads());
```

Comment:
We should not do any important work in destructor, or call any function that can throw exception. Does destructor in multiple threads takes significantly less time ?

Reply:
In case of huge hashtables - yes.
And right now the dtor can indeed throw, will rewrite to trySchedule and sequential destroy if thread allocation failed.

Reply:
If LOG_TRACE can throw exception, we should not write logs here. Not sure why we need logs in destructor.

Reply:
>If LOG_TRACE can throw exception

Likely it should not, since there is try/catch in `OwnSplitChannel::tryLogSplit` - https://github.com/ClickHouse/ClickHouse/blob/2482acc6004259d45ac1d0f009de176745087209/src/Loggers/OwnSplitChannel.cpp#L51-L76

> we should not write logs here. Not sure why we need logs in destructor.

To see how long does it takes and now it also has some information about how destroy goes.

Do you strongly against this? Or you want `try`/`catch`?

P.S. there are other places with logging in dtor.

Reply:
Reverted the change with trySchedule and just add a try/catch into dtor.